packages/gateway/src/durable-objects/gateway-rate-limits.js

```diff
@@ -0,0 +1,100 @@
+/**
+ * @typedef {Object} RateLimitResponse
+ * @property {boolean} shouldBlock request should be blocked
+ *
+ * @typedef {Object} RateLimitCharacteristics
+ * @property {number} RATE_LIMIT_REQUESTS
+ * @property {number} RATE_LIMIT_TIMEFRAME
+ */
+
+/**
+ * Durable Object to keep track of gateway rating limits.
+ * State: number[]
+ */
+export class GatewayRateLimits0 {
+  constructor(state) {
+    this.state = state
+    this.id = this.state.id.name
+    /** @type {RateLimitCharacteristics} */
+    this.rateLimitCharacteristics = getRateLimitingCharacteristics(this.id)
+
+    // `blockConcurrencyWhile()` ensures no requests are delivered until initialization completes.
+    this.state.blockConcurrencyWhile(async () => {
+      // Get state and initialize if not existing with gateway state
+      /** @type {Array<number>} */
+      this.rateLimitUsage = (await this.state.storage.get(this.id)) || []
+    })
+  }
+
+  // Handle HTTP requests from clients.
+  async fetch(request) {
+    // Apply requested action.
+    let url = new URL(request.url)
+    switch (url.pathname) {
+      case '/request':
+        // Filter out outdated requests
+        const now = Date.now()
+        this.rateLimitUsage = this.rateLimitUsage.filter(
+          (ts) => now - ts <= this.rateLimitCharacteristics.RATE_LIMIT_TIMEFRAME
+        )
+
+        // Add new request to track
+        if (
+          this.rateLimitUsage.length <
+          this.rateLimitCharacteristics.RATE_LIMIT_REQUESTS
+        ) {
+          this.rateLimitUsage.push(Date.now())
+          await this.state.storage.put(this.id, this.rateLimitUsage)
+
+          return new Response(
+            JSON.stringify({
+              shouldBlock: false,
+            })
+          )
+        }
+
+        // Block to prevent rate limit
+        return new Response(
+          JSON.stringify({
+            shouldBlock: true,
+          })
+        )
+      default:
+        return new Response('Not found', { status: 404 })
+    }
+  }
+}
+
+const MINUTE = 1000 * 60
+
+/**
+ * Get rate limiting characteristics of a given Gateway.
+ *
+ * @param {string} gatewayUrl
+ * @return {RateLimitCharacteristics}
+ */
+function getRateLimitingCharacteristics(gatewayUrl) {
+  switch (gatewayUrl) {
+    case 'https://ipfs.io':
+      return Object.freeze({
+        RATE_LIMIT_REQUESTS: 800,
```

Comment:
This one is quite difficult to obtain, considering the infrastructure setup.

ipfs.io public gateway uses several different techniques on load balancing and rate limiting, which makes it difficult to predict the actual rate limit value.

ipfs.io gateway starts by geo routing requests, followed by load balancing. The rate limit is set per load balancer. For instance, same {IP_ADDR, URI} are limited at1/second or 15/minute to a particular load balancer. Then there is a global 800/s on each Load Balancer.

Moreover, it also uses [bursting techniques](https://nginx.org/en/docs/http/ngx_http_limit_req_module.html#limit_req) that will start by delaying responses before it actually fails.

Reply:
I am thinking on performing the load tests once we have this new set of PRs merged in and see if we end up having any gateways rate limited. If so, try to tweak a bit the numbers